src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverter.cs

```diff
@@ -173,7 +173,11 @@ internal JsonConverter<TTarget> CreateCastingConverter<TTarget>()
 
         internal static bool ShouldFlush(Utf8JsonWriter writer, ref WriteStack state)
         {
-            // If surpassed flush threshold then return false which will flush stream.
+            // If surpassed flush threshold then return true which will flush stream.
+            if (state.PipeWriter is { CanGetUnflushedBytes: true } pipeWriter)
+            {
+                return state.FlushThreshold > 0 && pipeWriter.UnflushedBytes > state.FlushThreshold;
+            }
             return (state.FlushThreshold > 0 && writer.BytesPending > state.FlushThreshold);
```

Comment:
Do we still need that second branch? My impression is that all async serializers now write to a `PipeWriter`, so this could just return false I think.